fastcache_flush.go

```diff
@@ -0,0 +1,230 @@
+
+func (b *bucket) onFlushTick(flushInterval int64) {
+	if b.flusher.count > 0 && time.Since(time.UnixMilli(b.latestTimestamp)).Milliseconds() >= flushInterval {
+		b.setBatch(b.flusher)
+		b.cleanFlusher(b.flusher)
+	}
+}
+
+func (b *bucket) onNewItem(i *queuedStruct, maxBatch int, flushInterval int64) {
+	defer releaseQueuedStruct(i)
+
+	f := b.flusher
+
+	index := b.flusher.index.Load().([]flushChunkIndexItem)
+	indexId := i.h % uint64(len(index))
+	duplicated := index[indexId].exists(i.h)
+	forceFlush := false
+	if !duplicated {
+		kvLength := uint64(4) + uint64(len(i.K)) + uint64(len(i.V))
+		idxNew, newChunk := f.incrementIndexes(kvLength)
+		if newChunk {
+			f.currentFlushChunkId++
+		}
+		if f.currentFlushChunkId <= int32(len(f.chunks)) {
+			flushChunk := &f.chunks[f.currentFlushChunkId]
+
+			lenBuf := makeKvLenBuf(i.K, i.V)
+			//f.spinlock.Lock()
+			copy(flushChunk.chunk[flushChunk.size:], lenBuf[:])
+			copy(flushChunk.chunk[flushChunk.size+4:], i.K)
+			copy(flushChunk.chunk[flushChunk.size+4+uint64(len(i.K)):], i.V)
+
+			//f.spinlock.Unlock()
+			flushChunk.h = append(flushChunk.h, i.h)
+			flushChunk.idx = append(flushChunk.idx, f.idx)
+			flushChunk.gen = append(flushChunk.gen, f.gen)
+			flushChunk.chunkId = f.currentChunkId
+			flushChunk.cleanChunk = newChunk
+
+			for j := range index[indexId].h {
+				if atomic.LoadUint64(&index[indexId].h[j]) == 0 {
+					atomic.StoreUint64(&index[indexId].currentIdx[j], flushChunk.size)
+					atomic.StoreInt32(&index[indexId].flushChunk[j], f.currentFlushChunkId)
+					atomic.StoreUint64(&index[indexId].h[j], i.h)
+					break
+				}
+			}
+
+			b.flusher.index.Store(index)
+
+			flushChunk.size += kvLength
+			f.idx = idxNew
+			f.count++
+			if b.latestTimestamp == 0 {
+				b.latestTimestamp = time.Now().UnixMilli()
+			}
+			b.flusher.chunkSynced.Store(b.flusher.chunks)
+			atomic.AddUint64(&b.writeBufferSize, 1)
+		} else {
+			atomic.AddUint64(&b.droppedWrites, 1)
+			forceFlush = true
+		}
+
+	} else {
+		atomic.AddUint64(&b.duplicatedCount, 1)
+	}
+
+	if f.count >= maxBatch || (f.count > 0 && (time.Since(time.UnixMilli(b.latestTimestamp)).Milliseconds() >= flushInterval || forceFlush)) {
+		b.setBatch(f)
+		b.cleanFlusher(f)
+	}
+}
+
+func (b *bucket) setBatch(f *flusher) {
+	atomic.AddUint64(&b.setCalls, uint64(f.count))
+	atomic.AddUint64(&b.batchSetCalls, 1)
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	for i := int32(0); i <= f.currentFlushChunkId; i++ {
+		f := &f.chunks[i]
+		chunk := b.chunks[f.chunkId]
+		if chunk == nil {
+			chunk = getChunk()[:0]
+		} else if f.cleanChunk {
+			chunk = chunk[:0]
+		}
+
+		b.chunks[f.chunkId] = append(chunk, f.chunk[:f.size]...)
+
+		for j := 0; j < len(f.h); j++ {
+			b.m[f.h[j]] = f.idx[j] | (f.gen[j] << bucketSizeBits)
+		}
+	}
+	b.idx.Store(f.idx)
+	b.gen = f.gen
+	if f.needClean {
+		b.cleanLocked(f.idx)
+	}
+}
+
+func (b *bucket) cleanFlusher(f *flusher) {
+	index := b.flusher.index.Load().([]flushChunkIndexItem)
+	f.flushed.Store(true)
```

Comment:
I don't quite get this logic. So we take `flusher.index` first and then set `f.flushed` to `true`. But it means that we have race condition with `tryFindInFlushIndex`? That function is reading-only, but here we do some non-atomic stuff like 

```
if atomic.LoadUint64(&index[i].h[j]) != 0 {
				atomic.StoreUint64(&index[i].currentIdx[j], 0)
				atomic.StoreInt32(&index[i].flushChunk[j], 0)
				atomic.StoreUint64(&index[i].h[j], 0)
```

so I guess that function can be pretty much messed up